Refactor redundant code between standalone and QIIME 2 RPCA methods

fedarko opened the pull request:
- Add minimal required options (`--in_biom` and `--output_dir`).
- Try to create the specified output directory (if it already exists, this still works fine).
- A few misc. typo fixes
- Fix #28: abstract shared RPCA functionality to a new module, `deicode/rpca.py` (which contains `rpca()`, a function that runs RPCA given various parameters)
   - The QIIME 2 plugin setup directly references this module's function. From the perspective of a Q2 DEICODE plugin user, nothing should change.
   - The standalone DEICODE code also references this module's function. In order to make things a bit more consistent with the Q2 code, this includes **various minor breaking changes** (i.e. some commands have been renamed). To be exact:
     - `--in_biom` -> `--in-biom`
     - `--output_dir` -> `--output-dir`
     - `--min_sample_depth` -> `--min-sample-count`
   -  For the sake of clarity (and to avoid namespace conflicts), `deicode/scripts/_rpca.py` has been renamed to `deicode/scripts/_standalone_rpca.py`. I updated `deicode/scripts/__init__.py` and `setup.py` accordingly.
   - Also, the `--min-feature-count` and `--iterations` options have been added to the standalone code. This means that all of the functionality from the Q2 version of DEICODE should now be available in the standalone DEICODE version. (...this was the main reason I did this today :))
- Abstract shared RPCA parameter settings (descriptions and default values) to a new module, `deicode/_rpca_defaults.py`. This way, parameter settings only have to be modified once to take effect.
- I forgot to mention it in the commit message, but b2f7c4e98cc14fe3aa613a131677d001a2e30fc4 makes it so that the `show_default` Click argument is used for all the standalone RPCA parameters with default values.

fedarko committed:
ENH: Make minimal options required; fix some typos

Previously, running deicode standalone from the command line (i.e.
just "deicode") would give a confusing error due to the inputs being
of type None.

Marking certain options as "required" makes it clearer that the
user is missing something when they run DEICODE without specifying
an input BIOM and/or an output directory.

fedarko committed:
ENH: Standalone DEICODE: try creating output dir

IMO this makes it a bit easier to use standalone DEICODE. This is
sort of a subjective decision, so feel free to reject it if you
want :)

fedarko committed:
MAINT: fix some typos

fedarko committed:
MAINT: fix tiny typo in moving pictures tutorial

fedarko committed:
MAINT: Merge Q2 and non-Q2 RPCA code (fixes #28)

Now, the bulk of the RPCA code is just contained in deicode/rpca.py.
The standalone DEICODE code (deicode/scripts/_rpca.py) calls the
function in deicode/rpca.py, and does the extra work of loading the
BIOM table, writing the output files, etc. that Q2 normally takes care
of.

(The setup is somewhat similar to how rankratioviz uses generate.py
from both its Q2 and standalone codebases; a notable difference is
that we don't even have a Q2 _method file here anymore, since the
_method code is basically a subset of the code needed for the
standalone version.)

It's worth testing this a bit, but this should make maintaining
things a lot easier.

OTHER NOTES:

-This means that now all of the (previously exclusive) Q2 RPCA
 options are now available when running DEICODE outside of Q2. So now
 you can specify --minimum-feature-count and --iterations when running
 DEICODE outside of Q2.

-I changed some of the argument names/help messages to be consistent
 with Q2. THIS WILL BREAK prior code that uses DEICODE outside of
 Q2. (However, code that uses DEICODE within Q2 shouldn't be
 affected.)

-IMO, there is still some non-necessary redundancy in 1) the help
 messages and 2) the default settings of the various RPCA options.
 It may be worth defining a module where all of this information is
 stored to further eliminate redundancy (but we're already doing a lot
 better on that front than previously).

fedarko committed:
BUG: Rename _rpca to _standalone_rpca

fixes a bug in the refactoring done in the prior commit. Now,
all the tests pass.

Potentially worth noting that the test output (at least for the
standalone RPCA) looks slightly different from what's in the
repository. However, it looks like the values are only slightly
different (it's probably due to numpy version diffs/machine precision
diffs or something), and all the tests pass now, so I think this
should be ok (but this isn't my call to make :)).

At this point, I think this branch might be ready to merge back
into biocore. But I'd like to see what y'all have to say about it.

fedarko committed:
MAINT: Create and integrate _rpca_defaults module

This centralizes the various default parameters for the RPCA options,
cutting down on redundancy. Each option should only have to be changed
once now (so the Q2 and standalone DEICODE versions should be a lot
more consistent).

fedarko committed:
MAINT/BUG: actually add in _rpca_defaults module

(forgot to add it in the last commit, whoops)

fedarko committed:
MAINT: Update setup standalone entry point

fedarko committed:
MAINT: Merge param. descriptions for RPCA methods

This should pretty much knock out #28. At this point, pretty much
all of the standalone RPCA code is needed specifically for running
DEICODE outside of QIIME 2: that is, there's minimal redundancy
between the standalone and Q2 RPCA logic.

fedarko committed:
Merge pull request #1 from fedarko/merge_redundant_code

Merge redundant code

fedarko commented:
Looks like having less code decreased the overall coverage slightly. See https://github.com/lemurheavy/coveralls-public/issues/565#issuecomment-120918416 for reference.

cameronmartino reviewed (commented):
This is awesome, thank you @fedarko so much for this contribution. 

I only have one request to increase the version number and update the changelog. Then I think this is ready to merge!

fedarko commented:
Got it, changelog and version bump should be ready later today. Does `0.1.7` sound good as a new version number? Since some of these changes aren't backwards compatible, not sure if e.g. `0.2.0` might be a better option. (Your call, of course :))

fedarko committed:
REL: Update changelog; bump version to 0.1.7 #29

fedarko committed:
STY: make changelog a little prettier

Spaces out some bullet points; switch from the arrow syntax for name
changes to fancy-ish markdown tables

fedarko committed:
STY: oh geez attempt to fix changelog tables

fedarko committed:
STY: another stab at these tables

if this keeps not working i'll just revert to a prev commit

fedarko committed:
ENH: Update non-Q2 output filenames to match Q2

updated changelog accordingly

fedarko commented:
One more change -- the output filenames of the ordination and distance matrix produced by the standalone DEICODE RPCA now match the underlying filenames in the artifacts produced by the Q2 DEICODE RPCA. Updated the changelog accordingly.

ElDeveloper commented on CHANGELOG.md:
```diff
@@ -1,5 +1,95 @@
+| `--output_dir`       | `--output-dir`       |
+| `--min_sample_depth` | `--min-sample-count` |
+
+* `deicode.scripts._rpca` has been replaced by `deicode.scripts._standalone_rpca`.
```
This is rather minor but the convention is that objects prefixed with an underscore are deemed private, therefore I don't think you need to document changes to private functionality.  Though it is a nice thing to have.

fedarko replied:
Got it. I guess we can keep the references to "private" objects in the changelog for now.

ElDeveloper commented on CHANGELOG.md:
```diff
@@ -1,5 +1,95 @@
+  it didn't do before:
+  * Uses `--min-feature-count` with a default value of `10`. Previously, the
+    non-Q2 RPCA code didn't do this filtering step at all.
+  * Adds a PC3 containing zeros if the `rank` is set to `2`.
```
Is this because Emperor would fail otherwise? :O

fedarko replied:
@cameronmartino would probably know best! From the sounds of it (per his comment above), it sounds like this was because Emperor previously couldn't handle this (but it now can).

I actually removed this behavior from the RPCA code earlier today. Maybe it'd be a good idea (in the future) to add a test with `--rank 2` and seeing what happens when it's visualized with Emperor?

cameronmartino replied:
I believe this broke Emperor before the 2D option was available, that is why it is in there. I could be mistaken though.

ElDeveloper commented on CHANGELOG.md:
```diff
@@ -1,5 +1,95 @@
+### Miscellaneous
+
+* Since `deicode.rpca` is now used by both the QIIME 2 and non-QIIME 2 code,
+  the amount of redundant code has decreased. The primary benefit of this is
```
The last sentence in this paragraph is a bit odd, would you mind rephrasing?

fedarko replied:
Agreed, it does sound kind of awkward. Just pushed a commit that fixes this. Thanks!

ElDeveloper commented on CHANGELOG.md:
```diff
@@ -1,5 +1,95 @@
+| `Principal Coordinate Analysis` | `(Robust Aitchison) RPCA Biplot` |
+
+  * These changes shouldn't actually impact anything within the
+    `RPCA_Ordination.txt` files, since as of writing the [scikit-bio ordination format](http://scikit-bio.org/docs/latest/generated/skbio.io.format.ordination.html) doesn't include either of the method names. Just listing this here to be safe.
```
This would be reflected if interacting with the Python objects directly, so this is nice to document.

fedarko replied:
Sweet, good to know!

fedarko commented:
@cameronmartino Sorry to bother you with a ton of messages. I was looking through the code, and I don't think the standalone RPCA test (currently [located in test_rpca.py](https://github.com/biocore/DEICODE/blob/master/deicode/scripts/tests/test_rpca.py), but soon to be renamed to `test_standalone_rpca.py`) is correctly checking the output of anything. It looks like `dist_res`, `fea_res`, etc. will naturally be identical to `dist_exp`, `fea_exp`, etc. every time, since they're looking at the same exact files.

Furthermore, I'm somewhat confused about what `feature.txt`, `samples.txt`, etc. are supposed to be. It looks like these are earlier versions of the OrdinationResults output? Neither of these files -- nor `distance.txt` -- changes when the test is run. So, regardless of what the standalone RPCA code is doing, this test will always pass so long as it doesn't raise an error.

So: while we're at it with this PR, I think it would be worth **updating this test to look at the relevant files**: that is, verify that `distance-matrix.tsv` and `ordination.txt` (the new names for `RPCA_distance.txt` and `RPCA_Ordination.txt`) approximately match up with the ground truths.

It looks like `assert_array_almost_equal()` is still True when comparing `distance-matrix.tsv` and `truth_distance.txt`. However, I've been messing around with testing that the Ordination output matches `truth_feature.txt` and `truth_samples.txt` and it looks like **the signs of the first PC's values are flipped** between the "truth" feature and sample values and their respective rows in the Ordination. As an example, here's DEICODE's rank values for `OTU_2414` in the test data:

```OTU_2414	-4.094486487895422	2.7467056679841555	1.3098674678907734```

And here's the entry for this OTU in `truth_feature.txt`:

```OTU_2414	4.0930150209210066	2.749414394166293	1.3179536195044124```

PC2 and PC3's values are approximately equal (ish), but PC1 is flipped (if you negate the PC1 value, it's approximately equal to the other PC1 value). Not sure if this is an indication of a past bug that was fixed, or a bug that has been introduced since `truth_feature.txt` was created.

One possible reason why the values are only approximately equal (ignoring the "flipping" thing for now) is that, [as mentioned in the changelog](https://github.com/biocore/DEICODE/blob/d35b734078dd8db2c63cdf4ccfed1592faf2bdea/CHANGELOG.md#bug-fixes), this PR adds some of the stuff the Q2 RPCA was doing that the non-Q2 RPCA wasn't -- that is, adding a PC of 0s when `rank == 2`, and using `--min-feature-count` with a default value of `10`.

Also (not to get too nitpicky), but it looks like the [Q2 RPCA test](https://github.com/biocore/DEICODE/blob/master/deicode/q2/tests/test_method.py) doesn't looking any of the actual values of the OrdinationResults or distance matrix. Might be worth adding to that to ensure that these values make sense? Although I don't think I understand the math behind DEICODE well enough yet to do that.

fedarko committed:
TST: Make standalone RPCA test check correct files

It's failing, due to the "flipping" bug (?) and other small
differences between values as discussed in #29. The "truth" files
might just be really old.

I also added in "truth_sample_trimmed.txt", which is just a version
of the truth_sample.txt file without the OrdinationResults-ish stuff
at the end (so you can read it in pandas easier).

fedarko committed:
BLD: Remove sklearn dependency

Turns out scikit-learn just imports assert_array_almost_equal directly
from numpy, which is already a dependency of DEICODE. Seriously,
check it out!
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/testing.py

...so if nothing else installing DEICODE will now be a tiny bit faster
:)

fedarko committed:
REL: Small fix to changelog re new output filenms

cameronmartino commented:
@fedarko I apologize for the late response, thanks for fixing this. This was an oversight on my part.

1) Comparing files in the tests -- This makes perfect sense. The distance matrix should always be close and should not have negative inverses etc... The loadings, however, are probably only directly comparable by absolute values. This is because of the sample loadings and feature loadings can also be the inverse, everything is relative and directionality is not absolute in ordination (but one cannot be different and the other not change). Basically, if you multiply both sample/feature loadings in any of the "PCs" by negative one it won't change the real interpretation, you would just be flipping the numerator and denominator in the log-ratio. :) 

2) Editing tests -- It is definitely worth making all the tests check the real values, you are definitely right. As with above, it would probably be best to just check between the absolute values to some rounding error.  

 3) I think we can drop the adding zeros for rank < 3, that is updated in emperor now and will just default to 2D in that case (so that code is no longer needed). The Q2 script is definitely the most up to date script.

fedarko committed:
DEP: Don't add zeroes if rank < 3

Apparently this has been addressed in Emperor, per #29.

fedarko committed:
REL: Mention creating PC3 of 0s has been rm'd

fedarko committed:
REL: Reference DEICODE citation in Q2 plugin

Uses a BibTeX file from DEICODE's mSystems site
(https://msystems.asm.org/content/4/1/e00016-19), with the abstract
removed.

fedarko committed:
REL: Update changelog re: citation

fedarko committed:
TST: Let standalone RPCA test handle "flipping"

Still doesn't pass -- see #29 for details.

Also rewrote added test code a bit to be somewhat nicer.

fedarko commented:
@cameronmartino No worries, totally fine! Sorry for only noticing this stuff now while you're away. To respond to your points:

1. The explanation for some of the PC loadings "flipping" makes sense to me. I hadn't thought of it, but I can see how the ordination would still be interpreted basically the same way.
    - One thought: instead of just comparing the absolute values, it might be better to try checking each result PC against the expected PC, and if that fails then try negating one of the PCs and trying again. This way, if only some but not all of the loadings in a PC are flipped (which would be a problem), it'd still be caught. I mean that's probably not a super common error to happen, but IMO it's best to make the tests as robust as we can (within reason).
     - I pushed an update to the standalone RPCA test code that implements this idea, but it still fails because pandas says some of the values in one of the PCs differ (even with a large amount of leeway in the precision).
         - For reference: using `check_less_precise=2` for `pd.testing.assert_series_equal()`, pandas says the first feature PC values (between results and expected) differ by 25.65%. Decreasing `check_less_precise` to `1` lowers this difference to 3.15%, but at that point it's so imprecise that I'm not that convinced this test is really doing a lot.
         - If the `truth_feature.txt` file is just super old and not a reliable ground truth any more, it might be worth just using a new output file for the ordination test here.

2. Got it. This should be mostly the same as with the standalone test, right? I guess we can focus on fixing the standalone test first, and then apply the same logic to the Q2 test.

3. Understood. I've pushed a commit that removes this functionality from the RPCA code, and updated the changelog accordingly (now, instead of this functionality being *added* to the standalone DEICODE script, it's been *removed* from the Q2 DEICODE plugin).

Also @ElDeveloper thanks for the reviews! I'll update the changelog accordingly.

fedarko committed:
REL: Tidy up a sentence in the changelog #29

cameronmartino commented:
@fedarko 

1. I think that is a good approach!  I think the reason for the mismatch is, as you guessed they are not updated, probably spill over from not updating them in PR #27. I think it is fine just to update it. There is much more testing of the actual algorithm in tests/test_optspace.py. So this should just basically be wrapping that script. 

2. Yes, the exact same should apply to the Q2 tests. 

3. Awesome, thanks!

fedarko committed:
TST: Use new "truth" ordination for RPCA scrpt tst

Still uses the method discussed in #29, etc. that allows for PC
flipping.

fedarko committed:
STY: tiny fix to appease flake8

fedarko committed:
MAINT: Abstract OrdinationResults checking to func

fedarko committed:
TST: Test Q2 outputs against standalone outputs!

This should (keyword == should) finally address most of the concerns
raised in #29.

The committed code includes some commented-out functionality which
validates Q2 output against a ground truth, but since we're generating
the Q2 inputs randomly this doesn't really work. (Also, we already
nitpick the validity of the standalone DEICODE outputs in those
tests, and we check the standalone outputs against Q2 outputs here,
so... we should be fine. :)

I made some structural changes to the code while doing this, in
particular creating another unittest.TestCase subclass. This makes
sense to me personally because Testrpca only checks the rpca method
(not anything Q2-specific), while Test_qiime2_rpca actually checks
the Q2 RPCA method's implementation. That being said, if you'd like
to merge these classes feel free.

fedarko commented:
@cameronmartino Standalone test ground truths have been updated. I also added a Q2 test that runs the Q2 and standalone RPCA code on the same inputs and verifies their outputs are generally equal.

I think we're close to being done here. I just have one question left (super low priority): in `deicode/q2/tests/test_method`'s `test_rpca()` (which checks the basic validity of the rpca output objects), there are these two lines:
```python
        self.assertTrue(any(np.isnan(ord_test.features)))
        self.assertTrue(any(np.isnan(ord_test.samples)))
```
What's the purpose of the `any()` here? This seems to me like it should be saying "fail if there are any NaNs in either of these dataframes", but as far as I can tell this is actually saying "fail if there isn't at least one NaN in both of these dataframes" (which seems wrong, of course). The test is passing, so this isn't a problem or anything, but I'm just super confused about what this code is doing.

Anyway thanks for your advice! Excited to get these fixes pushed in next week :)

fedarko committed:
TST: Mark create_test_table() as not a test

cameronmartino commented:
@fedarko  Yes. you are right, nice catch! Now I am wondering how that is passing. I do not see any nan's when I test it by hand. Thanks so much for this contribution, this is looking great and much-needed update.

fedarko committed:
REL/TST: Add to changelog re: test code updates

Probably don't need to go into specifics here (documenting the
fixed test bugs is ok, but not sure we need to document just general
test updates)

fedarko committed:
TST: Properly check for NaN values in Ord. Results

fedarko commented:
Did some more digging, it looks like python's built-in `any()` and `all()` don't work as expected with pandas (see [here](https://stackoverflow.com/a/34273980/10730311) for details). It looks like they return True for inputs where they should return False -- here are the results from briefly testing this:

```python
>>> import pandas as pd
>>> df = pd.DataFrame({"A": [True, False], "B": [False, False]})
>>> df2 = pd.DataFrame({"A": [False, False], "B": [False, False]})
>>> any(df)
True
>>> any(df2)
True
>>> all(df)
True
>>> all(df2)
True
```

There are apparently pandas `.any()` and `.all()` DataFrame functions that actually work as you'd expect:
```python
>>> df.any(axis=None)
True
>>> df2.any(axis=None)
False
>>> df.all(axis=None)
False
>>> df2.all(axis=None)
False
```

Just pushed a fix for this. I think we're done, then?

Thanks!

fedarko committed:
REL: Update changelog re: date + q2 test bugfix

cameronmartino commented:
@fedarko  Tricky! This is good to know, thanks for teasing that apart. Will have to be careful with those data frames and missing values. I think we are good to go, feel free to merge when you are ready. Thanks!

fedarko commented:
@cameronmartino No problem, happy to help. I don't think I can merge this in since I don't have "write access," though.

fedarko committed:
REL: Make version 0.2.0; add non-Q2 docs to README

fedarko committed:
ENH/BUG: Add parameter sanity checking

For --rank and --iterations

fedarko committed:
DOC: Update README re: recently added minimum vals

fedarko commented:
Ok, so four things.

I finished up a test that generates an Emperor biplot with rank 2, as @cameronmartino and I discussed today. For reference, here's the relevant function (designed to be chucked into the `Test_qiime2_rpca` class in the DEICODE Q2 tests file, after making a few imports/etc.):

```python
    def test_rank2_matrix(self):
        """Verifies that using a rank 2 matrix works in DEICODE and Emperor."""

        ordination_qza, _ = q2deicode.actions.rpca(self.q2table, rank=2)
        # Generate fake sample metadata because it's required
        # self.table_sample_ids is just a list of the IDs in self.q2table
        fake_metadata_column_of_0s = [0] * len(self.table_sample_ids)
        sample_metadata_df = pd.DataFrame({"blah": fake_metadata_column_of_0s},
                                          index=self.table_sample_ids)
        sample_metadata_df.index.name = "Sample ID"
        # "Metadata" is imported from qiime2
        sample_metadata = Metadata(sample_metadata_df)
        biplot = q2emperor.actions.biplot(ordination_qza, sample_metadata)
        biplot.visualization.save("biplot.qzv")
```

I noticed a few things while writing/testing this code:

1. DEICODE didn't really check the rank or iterations parameters' values up front, so for the sake of clarity I added some basic parameter checking in https://github.com/biocore/DEICODE/pull/29/commits/140bd75c99be793ce446e9699e8d408e2ee4cac0 (where it ensures that `--rank` is at least `2`, and that `--iterations` is at least `1`). (Previously, using a rank of 0 or less would still give you an error from `scipy.sparse.linalg.svds()`, but the error message wasn't super clear and this still allowed a rank of 1. So I figure this is a beneficial change.)

2. q2-emperor can produce a biplot from a rank 2 ordination (as is generated in the new test code), but **it fails when loading the resulting QZV files in the browser** (checking the developer console, I get the same error messages* as I do when trying to load `bray_curtis_emperor.qzv` from [this forum post](https://forum.qiime2.org/t/problem-viewing-emperor-plot-with-qiime-tools-view-command-and-view-qiime2-org/5588/3)). This occurs with Chrome 73.0.3683.86 and QIIME 2 2019.1.

   I was under the impression that https://github.com/biocore/emperor/commit/a93f029548c421cb0ba365b4294f7a5a6b0209ce had fixed this problem in Emperor. Maybe this just wasn't included in the q2-emperor 2019.1 release? (If needed, we can just add back the "add a PC of zeroes" approach to DEICODE in the meantime.)

   For reference: [here's a link](https://drive.google.com/file/d/1Her9IViOKzXqFJ-RbqN8yZKf6HOXW-Cl/view?usp=sharing) to the biplot QZV file produced by the test code at the top of this post (sorry, github doesn't let me upload QZVs for some reason). And for comparison, [here's](https://drive.google.com/file/d/1XSYaWaxrrGZYAXaZy4JjJq9NcsKe7Swp/view?usp=sharing) a link to the biplot QZV file produced with the same code but using `rank=3` in the call to `rpca()` instead of `rank=2` (just to show that the `rank=2` is the problem here).

   <sup>*with slightly different line numbers, for some reason? I think this is due to Emperor version differences within the forum's QZV file, since I'm using Q2 2019.1 and that post was from 2018.</sup>

3. Since Emperor is failing in the browser, I'm not sure that a test in DEICODE is the best way to check this error (it seems like it'd be easier to do the testing in Emperor's repository's JS tests, if anything). This is why I haven't committed the test code in this post yet.

4. This isn't related to points 2 and 3 in this post, but I think I might have noticed a bug in `deicode/_optspace.py` (see [here](https://github.com/biocore/DEICODE/blob/master/deicode/_optspace.py#L55)). Since the code uses `1` as a starting point in the iteration loop -- and since it assigns to index `i+1` in the `dist` ndarray -- `dist` will always have a value of `0` at its second position (index `1`), regardless of `niter` (which is just `--iterations`). This also means that the loop's body won't execute when `--iterations` is 1, since `len(range(1, 1)) == 0`.

   You can test this out by adding a line that says `print(dist)` right after the for loop I linked to. I'm not familiar enough with the OptSpace algorithm to know the best way to fix this (or what its impacts on prior DEICODE output might have been), but this seems like unintentional behavior to me.


Anyway, that's what I've found so far. Sorry for the walls of text. @cameronmartino @ElDeveloper maybe we can meet sometime tomorrow / later this week to go over some or all of these points? I'm heading home now but I can try to help out with finishing this PR / squashing the relevant bugs this week. Thanks!

cameronmartino committed:
add iteration plus one in OptSpace to fix indexing issue, update test in scripts now off by third dec.

cameronmartino commented:
@fedarko Nice catch, I just pushed the iteration in the OptSpace helper by one. It should have little to no effect because of one iteration won't affect much once it has converged. I think in the test it changed in the third or fourth decimal, I updated the test to reflect this.

cameronmartino commented:
@fedarko (**1**) Could we move this into the actual optspace.py function. Maybe here: 

https://github.com/biocore/DEICODE/blob/475cfd581c811312d5793b4257c811ef6ba59e03/deicode/optspace.py#L130

cameronmartino committed:
moved rank<2 & iteration<1 value_errors to optspace along with docstring. Also added rank==2 addition of zeros in the PC3 to prevent breaking emperor

cameronmartino commented:
@fedarko (**2/3**) I added the zeros back for now, this also allows us some ability to be backcompatiable with older versions of Emperor. I think what is breaking this is that there are only two axis given. When even in the 2D case emperor seems to expect to see more than two axis in the Ordination file.  Unless @ElDeveloper objects to this work around, I will keep it.

cameronmartino committed:
removed rank<2 & iteration<1 value_errors to optspace along with docstring

cameronmartino committed:
Add more tests to OptSpace, Remove Inverseclr (this is already implemented in skbio), Remove coverge blocks on code and add tests for them

ElDeveloper commented:
I believe the workaround is OK to keep.

On (Apr-09-19|10:45), Cameron Martino wrote:
>@fedarko (**2/3**) I added the zeros back for now, this also allows us some ability to be backcompatiable with older versions of Emperor. I think what is breaking this is that there are only two axis given. When even in the 2D case emperor seems to expect to see more than two axis in the Ordination file.  Unless @ElDeveloper objects to this work around, I will keep it.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/biocore/DEICODE/pull/29#issuecomment-481359653

cameronmartino committed:
remove ratioviz function, reflect chnages in the chnagelog

cameronmartino committed:
remove ratioviz function, reflect chnages in the chnagelog

fedarko committed:
REL: Mention using PC3 of 0s in CHANGELOG

This adds back the text first added to the changelog in 5daf5de,
and removes the text added in ce05e00.

fedarko committed:
REL: make changelog a tiny bit prettier

cameronmartino committed:
update changelog, fix recursion issue with duplicates issue #30

cameronmartino committed:
Merge branch 'master' of https://github.com/fedarko/DEICODE

fedarko committed:
TST: Add Makefile with "make test"

Should make testing easier

fedarko committed:
Merge branch 'master' of https://github.com/fedarko/DEICODE

asdofijosdifjoaisdjfoaidsj

fedarko committed:
REL: tiny wording fix in the changelog

cameronmartino commented:
@fedarko Thanks again for this contribution. Will merge and update conda today.
